hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/WALEntryStream.java

```diff
@@ -206,73 +206,125 @@ private void setCurrentPath(Path path) {
+    boolean beingWritten = walFileLengthProvider.getLogFileSizeIfBeingWritten(nextPath).isPresent();
+    LOG.debug("Creating new reader {}, startPosition={}, beingWritten={}", nextPath,
+      currentPositionOfEntry, beingWritten);
+    try {
+      reader = WALFactory.createTailingReader(fs, nextPath, conf,
+        currentPositionOfEntry > 0 ? currentPositionOfEntry : -1);
+      return HasNext.YES;
+    } catch (WALHeaderEOFException e) {
+      if (!eofAutoRecovery) {
+        // if we do not enable EOF auto recovery, just let the upper layer retry
+        // the replication will be stuck usually, and need to be fixed manually
         return HasNext.RETRY;
       }
+      // we hit EOF while reading the WAL header, usually this means we can just skip over this
+      // file, but we need to be careful that whether this file is still being written, if so we
+      // should retry instead of skipping.
+      LOG.warn("EOF while trying to open WAL reader for path: {}, startPosition={}", nextPath,
+        currentPositionOfEntry, e);
+      if (beingWritten) {
+        // just retry as the file is still being written, maybe next time we could read
+        // something
+        return HasNext.RETRY;
+      } else {
+        // the file is not being written so we are safe to just skip over it
+        dequeueCurrentLog();
+        return HasNext.RETRY_IMMEDIATELY;
+      }
+    } catch (LeaseNotRecoveredException e) {
+      // HBASE-15019 the WAL was not closed due to some hiccup.
+      LOG.warn("Try to recover the WAL lease " + nextPath, e);
+      AbstractFSWALProvider.recoverLease(conf, nextPath);
+      return HasNext.RETRY;
+    } catch (IOException | NullPointerException e) {
+      // For why we need to catch NPE here, see HDFS-4380 for more details
+      LOG.warn("Failed to open WAL reader for path: {}", nextPath, e);
+      return HasNext.RETRY;
+    }
+  }
+
+  private HasNext lastAttempt() {
+    LOG.debug("Reset reader {} for the last time to pos {}, reset compression={}", currentPath,
+      currentPositionOfEntry, state.resetCompression());
+    try {
+      reader.resetTo(currentPositionOfEntry, state.resetCompression());
+    } catch (IOException e) {
+      LOG.warn("Failed to reset reader {} to pos {}, reset compression={}", currentPath,
+        currentPositionOfEntry, state.resetCompression(), e);
+      // just leave the state as is, next time we will try to reset it again, but there is a
+      // nasty problem is that, we will still reach here finally and try reset again to see if
+      // the log has been fully replicated, which is redundant, can be optimized later
+      return HasNext.RETRY;
+    }
+    Pair<WALTailingReader.State, Boolean> pair = readNextEntryAndRecordReaderPosition();
+    state = pair.getFirst();
+    // should not be written
+    assert !pair.getSecond();
+    if (!state.eof()) {
+      // we still have something to read after reopen
+      return HasNext.YES;
```

Comment:
`// just sleep a bit and retry to see if there are new entries coming since the file is
// still being written
return HasNext.RETRY;"`
When state not eof, the return value changed from HasNext.RETRY to HasNext.YES?

Reply:
This should also be a problem in the old code. We should return something here but in the old code, we just fall into the next switch section...
And after consideration, I think here we also need to consider whether there are errors, let me update the code.

Reply:
Updated. Will return RETRY if there are errors, otherwise return YES. If still get EOF, we can dequeue the current file.